引入依赖后，导致我项目的Jackson出现转换异常，去掉依赖后正常

2025-07-22 16:21:51.717 ERROR --- [nio-9999-exec-4] c.k.i.handler.GlobalExceptionHandler     L131 : [系统异常]Type definition error: [simple type, class java.time.LocalDateTime]
org.springframework.http.converter.HttpMessageConversionException: Type definition error: [simple type, class java.time.LocalDateTime]
	at org.springframework.http.converter.json.AbstractJackson2HttpMessageConverter.writeInternal(AbstractJackson2HttpMessageConverter.java:489)
	at org.springframework.http.converter.AbstractGenericHttpMessageConverter.write(AbstractGenericHttpMessageConverter.java:114)
	at org.springframework.web.servlet.mvc.method.annotation.AbstractMessageConverterMethodProcessor.writeWithMessageConverters(AbstractMessageConverterMethodProcessor.java:297)
	at org.springframework.web.servlet.mvc.method.annotation.RequestResponseBodyMethodProcessor.handleReturnValue(RequestResponseBodyMethodProcessor.java:192)
	at org.springframework.web.method.support.HandlerMethodReturnValueHandlerComposite.handleReturnValue(HandlerMethodReturnValueHandlerComposite.java:78)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:136)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:926)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:831)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1089)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:979)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1014)
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:903)
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:564)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:885)
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:658)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:195)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:51)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at com.alibaba.druid.support.jakarta.WebStatFilter.doFilter(WebStatFilter.java:113)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at cn.dev33.satoken.filter.SaPathCheckFilterForJakartaServlet.doFilter(SaPathCheckFilterForJakartaServlet.java:55)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:167)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:90)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:483)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:115)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:93)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:344)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:389)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:63)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:904)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1741)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1190)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:63)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Java 8 date/time type `java.time.LocalDateTime` not supported by default: add Module "com.fasterxml.jackson.datatype:jackson-datatype-jsr310" to enable handling (through reference chain: com.king.common.api.Result["data"]->com.king.modules.system.domain.vo.SysUserVO["createTime"])
	at com.fasterxml.jackson.databind.exc.InvalidDefinitionException.from(InvalidDefinitionException.java:77)
	at com.fasterxml.jackson.databind.SerializerProvider.reportBadDefinition(SerializerProvider.java:1330)
	at com.fasterxml.jackson.databind.ser.impl.UnsupportedTypeSerializer.serialize(UnsupportedTypeSerializer.java:35)
	at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:732)
	at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:770)
	at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:183)
	at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:732)
	at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:770)
	at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:183)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:502)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:341)
	at com.fasterxml.jackson.databind.ObjectWriter$Prefetch.serialize(ObjectWriter.java:1574)
	at com.fasterxml.jackson.databind.ObjectWriter.writeValue(ObjectWriter.java:1061)
	at org.springframework.http.converter.json.AbstractJackson2HttpMessageConverter.writeInternal(AbstractJackson2HttpMessageConverter.java:483)
	... 54 common frames omitted

--- a/dify/dify-client/dify-client-spring6/src/main/java/io/github/guoshiqiufeng/dify/client/spring6/utils/DifyMessageConverters.java
+++ b/dify/dify-client/dify-client-spring6/src/main/java/io/github/guoshiqiufeng/dify/client/spring6/utils/DifyMessageConverters.java
@@ -18,9 +18,11 @@ package io.github.guoshiqiufeng.dify.client.spring6.utils;
 import com.fasterxml.jackson.annotation.JsonInclude;
 import com.fasterxml.jackson.databind.DeserializationFeature;
 import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializationFeature;
 import org.springframework.http.converter.json.MappingJackson2HttpMessageConverter;
 import org.springframework.web.client.RestClient;
 
+import java.util.Optional;
 import java.util.function.Consumer;
 
 /**
@@ -34,16 +36,34 @@ public class DifyMessageConverters {
     }
 
     public static Consumer<RestClient.Builder> messageConvertersConsumer() {
-        return builder ->
-                builder.messageConverters(ms -> ms.stream()
+        return builder -> {
+            builder.messageConverters(converters -> {
+                // 找到现有的Jackson converter
+                Optional<MappingJackson2HttpMessageConverter> existingConverter = converters.stream()
                         .filter(c -> c instanceof MappingJackson2HttpMessageConverter)
-                        .findFirst()
-                        .ifPresent(converter -> {
-                            MappingJackson2HttpMessageConverter jacksonConverter = (MappingJackson2HttpMessageConverter) converter;
-                            ObjectMapper objectMapper = new ObjectMapper();
-                            objectMapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);
-                            objectMapper.disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
-                            jacksonConverter.setObjectMapper(objectMapper);
-                        }));
+                        .map(c -> (MappingJackson2HttpMessageConverter) c)
+                        .findFirst();
+
+                // 创建新的converter，基于现有配置
+                MappingJackson2HttpMessageConverter newConverter;
+                if (existingConverter.isPresent()) {
+                    ObjectMapper newMapper = existingConverter.get().getObjectMapper().copy();
+                    newMapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);
+                    newMapper.disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
+                    newConverter = new MappingJackson2HttpMessageConverter(newMapper);
+                } else {
+                    // 如果没有找到，创建新的
+                    ObjectMapper objectMapper = new ObjectMapper();
+                    objectMapper.disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);
+                    objectMapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);
+                    objectMapper.disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
+                    newConverter = new MappingJackson2HttpMessageConverter(objectMapper);
+                }
+
+                // 移除旧的，添加新的
+                converters.removeIf(c -> c instanceof MappingJackson2HttpMessageConverter);
+                converters.add(0, newConverter);
+            });
+        };
     }
 }

--- a/dify/dify-client/dify-client-spring6/src/test/java/io/github/guoshiqiufeng/dify/client/spring6/utils/DifyMessageConvertersTest.java
+++ b/dify/dify-client/dify-client-spring6/src/test/java/io/github/guoshiqiufeng/dify/client/spring6/utils/DifyMessageConvertersTest.java
@@ -16,7 +16,9 @@
 package io.github.guoshiqiufeng.dify.client.spring6.utils;
 
 import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.databind.DeserializationFeature;
 import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializationFeature;
 import org.junit.jupiter.api.DisplayName;
 import org.junit.jupiter.api.Test;
 import org.mockito.ArgumentCaptor;
@@ -28,7 +30,7 @@ import java.util.ArrayList;
 import java.util.List;
 import java.util.function.Consumer;
 
-import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.*;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.when;
 
@@ -66,12 +68,74 @@ public class DifyMessageConvertersTest {
         ObjectMapper objectMapper = new ObjectMapper();
         MappingJackson2HttpMessageConverter jacksonConverter = new MappingJackson2HttpMessageConverter(objectMapper);
         converters.add(jacksonConverter);
+        
+        // Add another converter to verify ordering
+        HttpMessageConverter<?> otherConverter = mock(HttpMessageConverter.class);
+        converters.add(otherConverter);
 
         // Apply the captured consumer to our test list
         consumerCaptor.getValue().accept(converters);
 
-        // Verify that the ObjectMapper's serialization inclusion is set to NON_NULL
-        // Using the non-deprecated API in newer Jackson versions
-        assertEquals(JsonInclude.Include.USE_DEFAULTS, objectMapper.getSerializationConfig().getDefaultPropertyInclusion().getValueInclusion());
+        // Verify that the list now only has two converters (the original one was replaced)
+        assertEquals(2, converters.size());
+        
+        // Verify the first converter is a MappingJackson2HttpMessageConverter
+        assertTrue(converters.get(0) instanceof MappingJackson2HttpMessageConverter);
+        
+        // Get the ObjectMapper from the configured converter
+        ObjectMapper configuredMapper = ((MappingJackson2HttpMessageConverter) converters.get(0)).getObjectMapper();
+        
+        // Verify that the Jackson converter is properly configured
+        assertEquals(JsonInclude.Include.NON_NULL, configuredMapper.getSerializationConfig().getDefaultPropertyInclusion().getValueInclusion());
+        assertFalse(configuredMapper.isEnabled(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES));
+        
+        // Verify the second converter is still the other mock converter
+        assertSame(otherConverter, converters.get(1));
+    }
+    
+    @Test
+    @DisplayName("Test messageConvertersConsumer creates new converter when none exists")
+    public void testMessageConvertersConfigurationNoExistingConverter() {
+        // Create a consumer for message converters
+        Consumer<RestClient.Builder> consumer = DifyMessageConverters.messageConvertersConsumer();
+
+        // Create a mock RestClient.Builder
+        RestClient.Builder mockBuilder = mock(RestClient.Builder.class);
+
+        // Create an ArgumentCaptor to capture the Consumer<List<HttpMessageConverter<?>>> argument
+        @SuppressWarnings("unchecked")
+        ArgumentCaptor<Consumer<List<HttpMessageConverter<?>>>> consumerCaptor =
+                ArgumentCaptor.forClass(Consumer.class);
+
+        // Mock the behavior
+        when(mockBuilder.messageConverters(consumerCaptor.capture())).thenReturn(mockBuilder);
+
+        // Apply the consumer to the mock builder
+        consumer.accept(mockBuilder);
+
+        // Create a list without any Jackson converter
+        List<HttpMessageConverter<?>> converters = new ArrayList<>();
+        HttpMessageConverter<?> otherConverter = mock(HttpMessageConverter.class);
+        converters.add(otherConverter);
+
+        // Apply the captured consumer to our test list
+        consumerCaptor.getValue().accept(converters);
+
+        // Verify that the list now has two converters (added a new one)
+        assertEquals(2, converters.size());
+        
+        // Verify the first converter is a MappingJackson2HttpMessageConverter
+        assertTrue(converters.get(0) instanceof MappingJackson2HttpMessageConverter);
+        
+        // Get the ObjectMapper from the configured converter
+        ObjectMapper configuredMapper = ((MappingJackson2HttpMessageConverter) converters.get(0)).getObjectMapper();
+        
+        // Verify that the Jackson converter is properly configured
+        assertEquals(JsonInclude.Include.NON_NULL, configuredMapper.getSerializationConfig().getDefaultPropertyInclusion().getValueInclusion());
+        assertFalse(configuredMapper.isEnabled(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES));
+        assertFalse(configuredMapper.isEnabled(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS));
+        
+        // Verify the second converter is still the other mock converter
+        assertSame(otherConverter, converters.get(1));
     }
 }

